drivers: display: Add support for LED matrix driven by nRF SoC GPIOs

anangl opened the pull request:
Add a display driver and the corresponding devicetree binding for a LED
matrix with rows and columns driven by nRF SoCs GPIOs. Such matrix can
be found in the BBC micro:bit v2 board, so a proper devicetree node is
added to its definition so that the driver can be used for that board.

Add also a sample that could be used to see the nRF LED matrix
driver in action and that will prove that the driver is buildable for
both the bbc_microbit_v2 and bbc_microbit boards.

anangl committed:
drivers: display: Add support for LED matrix driven by nRF SoC GPIOs

Add a display driver and the corresponding devicetree binding for a LED
matrix with rows and columns driven by nRF SoCs GPIOs. Such matrix can
be found, for example, in the BBC micro:bit boards.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
samples: boards: nrf: Add sample for the display_nrf_led_matrix driver

Add a sample that can be used to see the nRF LED matrix display driver
in action and that will prove that the driver is buildable for both the
bbc_microbit_v2 and bbc_microbit boards.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

carlescufi commented:
@LingaoM @peterniebert @chudov this should be as generic as it gets.

gmarull commented on boards/arm/bbc_microbit_v2/bbc_microbit_v2.dts:
```diff
@@ -40,6 +40,31 @@
 		};
 	};
 
+	led_matrix {
```
nit: this addition should be part of a separate commit

gmarull commented on drivers/display/display_nrf_led_matrix.c:
```diff
@@ -0,0 +1,448 @@
+		     "Invalid column index in pixel-mapping["#idx"]");
+DT_FOREACH_PROP_ELEM(MATRIX_NODE, pixel_mapping, CHECK_PIXEL)
+
+#define BASE_FREQUENCY     16000000
+#define TIMER_CLK_CONFIG   NRF_TIMER_FREQ_16MHz
+#define PWM_CLK_CONFIG     NRF_PWM_CLK_16MHz
+#define REFRESH_FREQUENCY  50
+#define BRIGHTNESS_MAX     255
```
some of these parameters, e.g. `REFRESH_FREQUENCY` could be defined in DT.

anangl replied:
Added the `refresh-frequency` DT property.

gmarull commented on drivers/display/display_nrf_led_matrix.c:
```diff
@@ -0,0 +1,448 @@
+	bool    blanking;
+};
+
+static void set_pin(uint8_t pin_info, bool active)
+{
+	uint32_t value = active ? 1 : 0;
+
+	if (pin_info & ACTIVE_LOW_MASK) {
+		value = !value;
+	}
+	nrf_gpio_pin_write(pin_info & PSEL_MASK, value);
+}
```
does GPIO API add too much overhead?

anangl replied:
The overhead is not that important. The actual reason is that I anyway need to operate on PSEL values as I need to provide those in configurations of PWM and GPIOTE channels, see https://github.com/zephyrproject-rtos/zephyr/blob/c3dd5dc6071b6e82d5dbd6395b38e6bfb25911f5/drivers/display/display_nrf_led_matrix.c#L302-L310
And since GPIO interrupts are not used here, I see no reason to make this driver dependent on the GPIO one.

gmarull commented on drivers/display/display_nrf_led_matrix.c:
```diff
@@ -0,0 +1,448 @@
+
+static int api_blanking_on(const struct device *dev)
+{
+	struct display_drv_data *dev_data = dev->data;
+	const struct display_drv_config *dev_config = dev->config;
```
`data`, `config` are usual var names

anangl replied:
I find them too generic and I had cases in the past where some other `config` or `data` appeared in a given function. So to avoid confusion I'd prefer to keep these prefixes (these are only 4 extra characters). Many nRF drivers actually use such naming.

gmarull commented on drivers/display/display_nrf_led_matrix.c:
```diff
@@ -0,0 +1,448 @@
+	};
+
+	nrf_pwm_pins_set(dev_config->pwm, out_psels);
+	BUILD_ASSERT(PIXEL_PERIOD <= PWM_COUNTERTOP_COUNTERTOP_Msk);
```
suggestion: I'd centralize build asserts in a single place for better readability

gmarull commented on drivers/display/display_nrf_led_matrix.c:
```diff
@@ -0,0 +1,448 @@
+		uint8_t row_pin_info = dev_config->rows[i];
+
+		set_pin(row_pin_info, false);
+		nrf_gpio_cfg(row_pin_info & PSEL_MASK,
+			     NRF_GPIO_PIN_DIR_OUTPUT,
+			     NRF_GPIO_PIN_INPUT_DISCONNECT,
+			     NRF_GPIO_PIN_NOPULL,
+			     NRF_GPIO_PIN_S0S1,
+			     NRF_GPIO_PIN_NOSENSE);
+	}
+
+	for (uint8_t i = 0; i < COL_COUNT; ++i) {
+		uint8_t col_pin_info = dev_config->cols[i];
+
+		set_pin(col_pin_info, false);
+		nrf_gpio_cfg(col_pin_info & PSEL_MASK,
+			     NRF_GPIO_PIN_DIR_OUTPUT,
```
may GPIO API be used here?

gmarull commented on drivers/display/display_nrf_led_matrix.c:
```diff
@@ -0,0 +1,448 @@
+	.blanking   = true,
+};
+
+#define GET_PIN_INFO(node_id, pha, idx) \
+	(DT_GPIO_PIN_BY_IDX(node_id, pha, idx) | \
+	 (DT_PROP_BY_PHANDLE_IDX(node_id, pha, idx, port) << 5) | \
+	 ((DT_GPIO_FLAGS_BY_IDX(node_id, pha, idx) & GPIO_ACTIVE_LOW) ? \
+		ACTIVE_LOW_MASK : 0)),
```
ditto for GPIO stuff, gpio_dt_spec and GPIO API may be used

anangl replied:
See https://github.com/zephyrproject-rtos/zephyr/pull/40968#discussion_r771270367.

gmarull commented on dts/bindings/display/nrf-led-matrix.yaml:
```diff
@@ -0,0 +1,59 @@
+      description: |
+        Array of GPIOs to be used as columns of the matrix.
+
+    pixel-mapping:
```
This property could be optional on "regular" LED matrices (like the one in V2). I guess this exists due to the 3x9 matrix used as 5x5 in V1.

anangl replied:
Yes, for "regular" matrices, I could generate such array in the driver. But isn't it good to have the mapping specified explicitly in DTS so that everything regarding the hardware is clear?

gmarull replied:
fair enough

gmarull commented on samples/boards/nrf/nrf_led_matrix/boards/bbc_microbit.overlay:
```diff
@@ -0,0 +1,33 @@
+ */
+
+/ {
+	led_matrix {
```
can this be added to the board, as for v2?

anangl replied:
Done.

jfischer-no replied:
This overlay is no longer needed then.

anangl replied:
Yes, but I kept it is a template, as per:
https://github.com/zephyrproject-rtos/zephyr/blob/c3dd5dc6071b6e82d5dbd6395b38e6bfb25911f5/samples/boards/nrf/nrf_led_matrix/README.rst#L20-L21
That was the reason I haven't originally added the `led_matrix` node to the bbc_microbit board definition. Should I remove the overlay then and that sentence from README?

gmarull replied:
why not document it on the readme then? having an overlay is a bit confusing if not required.

gmarull replied:
I think the best place for that is the binding as an example (main description field)

anangl replied:
> why not document it on the readme then? having an overlay is a bit confusing if not required.

I agree, it's confusing. I'll remove the overlay.

gmarull commented on dts/bindings/display/nrf-led-matrix.yaml:
```diff
@@ -0,0 +1,59 @@
+
+description: Generic LED matrix driven by nRF SoC GPIOs
+
+compatible: "nrf-led-matrix"
```
should be `nrf,led-matrix`

anangl replied:
I used "nordic,nrf-led-matrix". The first part should be the vendor name, right?

gmarull replied:
correct

jfischer-no commented on samples/boards/nrf/nrf_led_matrix/src/main.c:
```diff
@@ -0,0 +1,197 @@
+	for (;;) {
+		printk("Started\n");
+
+		ret = display_set_brightness(dev, 0x7F);
+		if (ret < 0) {
+			printk("display_set_brightness failed: %u/%d\n",
+				__LINE__, ret);
+		}
+
+		buf[0] = 0xA8;
+		buf[1] = 0x40;
+		buf[2] = 0xA8;
+		buf[3] = 0x10;
+		buf[4] = 0xA8;
+		ret = display_write(dev, 0, 0, &buf_desc, buf);
+		if (ret < 0) {
+			printk("display_write failed: %u/%d\n",
+				__LINE__, ret);
+		}
+
+		ret = display_blanking_off(dev);
+		if (ret < 0) {
+			printk("display_blanking_off failed: %u/%d\n",
+				__LINE__, ret);
+		}
```
I suggest to move show blocks to own function led_matrix_show_foo(), led_matrix_show_bar()...

jfischer-no commented on drivers/display/display_nrf_led_matrix.c:
```diff
@@ -0,0 +1,448 @@
+{
+	caps->x_resolution = X_PIXELS;
+	caps->y_resolution = Y_PIXELS;
+	caps->supported_pixel_formats = PIXEL_FORMAT_MONO10;
```
PIXEL_FORMAT_MONO01, at least it should match one in api_set_pixel_format(), L:162

jfischer-no commented on drivers/display/display_nrf_led_matrix.c:
```diff
@@ -0,0 +1,448 @@
+DEVICE_DT_DEFINE(MATRIX_NODE,
+		 instance_init, NULL,
+		 &instance_data, &instance_config,
+		 APPLICATION, CONFIG_DISPLAY_INIT_PRIORITY, &driver_api);
```
`POST_KERNEL, CONFIG_DISPLAY_INIT_PRIORITY, &driver_api);`

anangl committed:
boards: bbc_microbit*: Add led_matrix devicetree node

Both BBC micro:bit v1 and v2 boards are equipped with a LED matrix
driven by the SoC GPIOs. Add "nordic,nrf-led-matrix" compatible nodes
to DTS files of both these boards so that the nRF LED matrix display
driver can be used for them.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl commented:
Rebased.

anangl commented:
Addressed review comments.

anangl commented:
The error and warnings reported by the compliance checks are false positives (in the comparisons, both sides are constant):
```
-:177: WARNING:CONSTANT_COMPARISON: Comparisons should place the constant on the right side of the test
#177: FILE: drivers/display/display_nrf_led_matrix.c:48:
+BUILD_ASSERT(PIXEL_PERIOD <= BIT_MASK(16));

-:179: WARNING:CONSTANT_COMPARISON: Comparisons should place the constant on the right side of the test
#179: FILE: drivers/display/display_nrf_led_matrix.c:50:
+BUILD_ASSERT(PIXEL_PERIOD <= PWM_COUNTERTOP_COUNTERTOP_Msk);

-:188: ERROR:SPACING: need consistent spacing around '*' (ctx:WxV)
#188: FILE: drivers/display/display_nrf_led_matrix.c:59:
+	NRF_PWM_Type *pwm;
```

gmarull commented on dts/bindings/display/nordic,nrf-led-matrix.yaml:
```diff
@@ -0,0 +1,66 @@
+    refresh-frequency:
+      type: int
+      required: false
+      default: 50
```
per policy: defaults need to be justified in the description

anangl replied:
I've made the property required instead.

gmarull commented on dts/bindings/display/nordic,nrf-led-matrix.yaml:
```diff
@@ -0,0 +1,66 @@
+
+    refresh-frequency:
+      type: int
+      required: false
```
```suggestion
```

zpm1066 commented:
@anangl Thank you.
